### DISCUSSION TITLE
Documentation about how the default dispatcher reuses connections

### DISCUSSION THREAD

**Discussion by brettwillis** [nodejs/undici#2382] [General]
When developing a high-performance server, one of the first things I'd do in `node-fetch` and similar was make sure I used a `new Agent({ keepAlive: true })` agent to make sure that I wasn't re-establishing and then immediately destroying fresh TCP connections for every request.

Now, particularly in the context of Node.js's built-in `fetch()`, my first question is: *if* and *for how long* are connections being kept alive and reused under the hood? Based on benchmarks, I'm guess it must be reusing connections in some way.

In a browser implementation of `fetch()`, this is purely up to the browser to optimise, and they are normally very good at pooling connections and even using pipelining and HTTP/2/3 etc.

In `undici`, these details are implemented by the dispatcher, and the in the case of build-in `fetch()` this is not something we can control or customise.

So my question/request is: can we have documentation that explains how the default dispatcher behaves in terms of reusing connections, opportunistically using pipelining and HTTP/2/3 etc.

Like I mentioned this is one of the first questions when developing a high-performance server, and currently we need to dig into the code to find there is a default dispatcher, and examine the implementation learn what it is doing.

**Comment by metcoder95** [comment]
Me again!

`fetch` uses the `globalDispatcher` set by `undici`. This dispatcher uses an [`Agent`](https://undici.nodejs.org/#/docs/api/Agent) with its defaults. Under-the-hood, `Agent` creates, depending on the configuration passed, an instance of either a [`Pool`](https://undici.nodejs.org/#/docs/api/Pool) or [`Client`](https://undici.nodejs.org/#/docs/api/Client), if entirely based on the defaults, a `Client` is made (it depends on the number of `connections` passed).

Usually, this configuration works perfectly for many cases, but for more complex use cases, you can alter it by passing an `Agent` instance to the `setGlobalDispatcher`, another option is to do the same but pass a `dispatcher-like` object to fetch (usually a Pool or Client instance), e.g. `fetch('url', { dispatcher })`.

H2 support (H3 is not in Node.js yet) is experimental, and can be used by creating a client or agent with the flag `allowH2: true`.

Hope is pretty clear, but a PR is welcome to improve our documentation around those topics 🙂

**Reply by longzheng** [reply]
In case anyone else was looking, the default agent checks for `connections === 1` https://github.com/nodejs/undici/blob/e93dbd59bd5b2db092f5ec3858b3b167288b4482/lib/dispatcher/agent.js#L19-L23

**Comment by Galienna** [comment]
I'm also struggling with the way Undici's Agent is managing its connections. With Node's Agent, we had a clear documentation [here](https://nodejs.org/docs/latest-v22.x/api/http.html#new-agentoptions), with `keepAlive`, `maxSockets` and especially `scheduling`.

Could you please explain what's the scheduling strategy here ? Is there a way ton configure it ?

**Reply by metcoder95** [reply]
Any contribution for enhancing the docs is really welcomed!

`Pool` uses a mix of round robin with resource based algorithm based on how busy the client is. `BalancedPool` on the other hand, weights based on connection stability and status of the pool of upstream connections.

**Reply by Galienna** [reply]
Thanks for the answer :) I need to understand the lib better before trying to enhance the doc :D
Maybe i can compile something after our discussion.

So, the Pool and its round robin sounds good to me. 
My use case is the following : I'm sending 5k req/s to a URL. The URL resolves through a GSLB with a short TTL (30s). The GSLB frequently withdraws one IP (out of 3) for an hour. Though i'm using `keepAlive`, I need to vary the IP i use to follow the load balancing planned by the GSLB. If i keep reusing the last connection, i may stick to an IP that has been withdrawn by the GSLB, isn't it ? Though i don't know how is managed the connection pool.

When sending a request to `mydns.com`, do you first resolve the DNS then try to find a connection that matches the given IP ? Or do you directly try to find a connection that has been used for `mydns.com` before resolving the DNS ?

**Reply by Galienna** [reply]
My tests seems to tell the workflow is :
1) `fetch('some URL')`
2) try to reuse a connection that matches the hostname (*without re-resolving hostname*)
3) if nothing found, resolve hostname and open a connection

So i'm sticking to the same connection and it fucks up load balancing and the whole purpose of the GSLB 😭 

The only thing that can save me on this one is the `Agent`'s option `maxRequestsPerClient`, that forces the connection to close after processing the desired number of requests.

Otherwise, i guess i can resolve the DNS myself to get fresh IP before calling `fetch()`

**Reply by metcoder95** [reply]
>When sending a request to mydns.com, do you first resolve the DNS then try to find a connection that matches the given IP ? Or do you directly try to find a connection that has been used for mydns.com before resolving the DNS ?

We leverage to `node:dns#lookup` for that as we use the `socket` defaults for IP resolution.
All pools are based on origins URLs, then whatever you've set as URL (either the direct IP or the DNS name), will be used to keep the Pool of connections for the downstream.

If the origin rotates IPs after a time, the connection will be dropped but a new one will be made automatically.

If seeking for a more customized behavior with DNS involvement, I'd recommend to use the `dns` interceptor, which caches and handles this automatically (but you can extend it if you need to).

**Reply by Galienna** [reply]
> If the origin rotates IPs after a time, the connection will be dropped but a new one will be made automatically.

Maybe i missed something, but my tests show that undici's fetch & agent are reusing an old connection with an IP that no longer matches the requested URL. I don't see the old connection being dropped.

My test is the following :
````javascript
// file test.js
// "undici": "6.21.0"
import undici from 'undici';

const testAgent = new undici.Agent({
    connections: 10,
    connect: {
        keepAlive: true,
        timeout: 5000
    }
});

async function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}

async function main() {
    while (1) {
        const response = await undici.fetch('https://some-partner.com', {
            dispatcher: testAgent,
            method: 'PATCH',
            body: '{"foo":"bar"}',
            headers: {'content-type': 'application/json'}
        });
        console.log('resp code '+response.statusText)
        
        await sleep(1000);
    }
}

main();
````
Then :
- In `/etc/hosts`, force the IP of `some-partner.com` to one of the IP provided by the GSLB : `10.117.30.10`
- Observe connections with `watch -n 0.3 "netstat -anotv | grep -E 'Adresse|443'"
- Run the test with node v20.18.0 : `node test.js`
- Netstat shows a connection : 
````
Proto Recv-Q Send-Q Adresse locale          Adresse distante        Etat       Timer
tcp        0      0 10.0.2.15:50376         10.117.30.10:443        ESTABLISHED keepalive (17,02/0/0)
````
- Edit `/etc/hosts` to simulate a rotation of IP. Set `some-partner.com` to : `10.117.28.10`
- Netstat shows the same connection being reused over and over, even if i keep running the test for 5 min :
````
Proto Recv-Q Send-Q Adresse locale          Adresse distante        Etat       Timer
tcp        0      0 10.0.2.15:50376         10.117.30.10:443        ESTABLISHED keepalive (54,50/0/0)
````

If you add a log to see what `dns.lookup` is getting, you'll see it properly follows `/etc/hosts`

**Reply by metcoder95** [reply]
Sorry, I didn't explain myself correctly; it is automatically dropped as soon as the current connection (before rotation) is ended by the downstream or the client itself. As long as the IP stills reachable, `undici` will keep the connection open.

Once the connection is closed, `undici` will attempt to create a new connection, and do a new dns lookup. Depending on the OS settings and the DNS records cached, it will resolve to the new rotated IP.

If you are looking to manage these resolutions in a more preemptive way, again, the best is to use the `dns` interceptor to customize the way the records are cached and how often can be rotated.

**Reply by Galienna** [reply]
Thank you for the detailed answer :) I'll try to figure out what is dns interceptors and how to use it with `fetch`. Just to be sure, you're talking about [this](https://undici.nodejs.org/#/docs/api/Dispatcher?id=dns), right ?
NB : the search tool from the doc gives nothing when searching for "interceptor" ^^'

That being said, i don't understand why the default behaviour of undici is to ignore DNS TTL, especially after reading your comment on [#3350](https://github.com/nodejs/undici/issues/3350#issuecomment-2301483505). I'm confused

**Reply by metcoder95** [reply]
>Thank you for the detailed answer :) I'll try to figure out what is dns interceptors and how to use it with fetch. Just to be sure, you're talking about [this](https://undici.nodejs.org/#/docs/api/Dispatcher?id=dns), right ?

Yes. And yeah, is quite hidden; sorry for that

>That being said, i don't understand why the default behaviour of undici is to ignore DNS TTL, especially after reading your comment on https://github.com/nodejs/undici/issues/3350#issuecomment-2301483505. I'm confused

It uses `node:dns`, so whatever the `dns.lookup` does, undici works with it. See [`dns` implementation considerations](https://nodejs.org/api/dns.html#implementation-considerations) for more info.
